File: agapay-web/src/repositories/stationRepository.ts
import { Station, computeAlertLevel } from '@/models/station';
import { TelemetryPoint, PredictionPoint } from '@/models/telemetry';
import { initialStations, generateTelemetryHistory, generateMLPredictions } from '@/services/mockData';

export interface StationRepository {
  getStations(): Promise<Station[]>;
  getStationById(id: string): Promise<Station | undefined>;
  getTelemetryHistory(stationId: string, hours: number): Promise<TelemetryPoint[]>;
  getMLPredictions(stationId: string): Promise<PredictionPoint[]>;
  updateStationThresholds(
    stationId: string,
    thresholds: { advisory: number; warning: number; evacuate: number }
  ): Promise<Station>;
}

export class MockStationRepository implements StationRepository {
  private stations: Station[] = [...initialStations];

  async getStations(): Promise<Station[]> {
    return [...this.stations];
  }

  async getStationById(id: string): Promise<Station | undefined> {
    return this.stations.find((s) => s.id === id);
  }

  async getTelemetryHistory(stationId: string, hours: number): Promise<TelemetryPoint[]> {
    return generateTelemetryHistory(stationId, hours);
  }

  async getMLPredictions(stationId: string): Promise<PredictionPoint[]> {
    return generateMLPredictions(stationId);
  }

  async updateStationThresholds(
    stationId: string,
    thresholds: { advisory: number; warning: number; evacuate: number }
  ): Promise<Station> {
    const index = this.stations.findIndex((s) => s.id === stationId);
    if (index === -1) throw new Error('Station not found');

    const updated: Station = {
      ...this.stations[index],
      advisoryThresholdCm: thresholds.advisory,
      warningThresholdCm: thresholds.warning,
      evacuateThresholdCm: thresholds.evacuate,
    };
    updated.alertLevel = computeAlertLevel(updated);
    this.stations[index] = updated;
    return updated;
  }
}

export const stationRepository = new MockStationRepository();

File: agapay-mobile/lib/repositories/station_repository.dart
import '../models/station.dart';
import '../models/telemetry_history.dart';
import '../services/mock_data.dart';

abstract class StationRepository {
  Future<List<Station>> getStations();
  Future<Station> getStationById(String id);
  Future<TelemetrySummary> getStationTelemetryHistory(String stationId, {int hoursBack = 24});
}

class MockStationRepository implements StationRepository {
  List<Station> _stations = List.from(MockData.initialStations);

  @override
  Future<List<Station>> getStations() async {
    await Future.delayed(const Duration(milliseconds: 250));
    return List.unmodifiable(_stations);
  }

  @override
  Future<Station> getStationById(String id) async {
    await Future.delayed(const Duration(milliseconds: 150));
    return _stations.firstWhere(
      (s) => s.id == id,
      orElse: () => _stations.first,
    );
  }

  @override
  Future<TelemetrySummary> getStationTelemetryHistory(String stationId, {int hoursBack = 24}) async {
    await Future.delayed(const Duration(milliseconds: 300));
    return MockData.getHistoricalSummary(stationId, hoursBack);
  }

  void updateStation(Station updated) {
    final index = _stations.indexWhere((s) => s.id == updated.id);
    if (index != -1) {
      _stations[index] = updated;
    }
  }
}
